fix: permission system blockers

bouassaba opened the pull request:
- fix: can give access to owner
- feat: group and organization membership are now based on permissions - not on derivable data (tables below)
- refactor: add migrations to remove the tables `organization_user` and `group_user`

Owner is now visible in the share dialog:

<img width="1410" src="https://github.com/user-attachments/assets/880b8b51-6583-4c46-9890-b987ed3e03a4">
<br/>
<br/>

Make sure at least one owner is available per organization to avoid dangling data:

<img width="1482" src="https://github.com/user-attachments/assets/50c5a71e-cd24-4b71-98ba-ded4cc954357">
<br/>
<br/>

Make sure at least one owner is available per group to avoid dangling data:

<img width="1436" src="https://github.com/user-attachments/assets/548989d2-2529-4979-b0fc-c68ab19ca030">
<br/>
<br/>

Closes #192

dsonck92 commented:
Wait, I thought we would rather switch it the other way around? Have membership carry the level of the membership? Or do you intend to reintroduce that down the line when we refactor the "multi foreign key" userpermission/grouppermission

bouassaba commented:
@dsonck92 What I did is I just fixed the bug, where a person couldn't grant permission to the owner, I didn't mean to introduce any new concept :) it's just a plain bug fix that can keep us running until you implement for us the new permission system.

Regarding the tables that I removed, it's because the previous logic was adding people to a table to make them members, and not adding the owner to that table, so the owner has the "owner" permission on the group/org but he is never on this table. So to solve this with the minimum change possible, the most logical thing seemed to get rid if these troublesome tables and say: "if a user has a permission on a group/org" it means he is a member".
As I said, this is just a quick "hack" until you introduce for us the new system.

bouassaba commented:
@dsonck92 I hope this clarifies the misunderstanding?

bouassaba commented:
@dsonck92 anyways, feel free to change my PR if you have a better idea ;) I really don't mind, I just wanted to help fix the issue temporarily.

bouassaba commented:
Just to be more clear about my opinion: for me I'm already well convinced that the current permission system has flaws, and it needs to be replaced with the new one, as soon as v3.0.
That's why while fixing this bug, I wanted to invest the minimum possible effort, because the current permission system is going to be thrown away, so it's wise to do just enough to allow "granting permissions to the owner" and unblock people who are using v2.0 until we give them v3.0.

dsonck92 commented:
@bouassaba Alright, clear, yes, so minimal changes to get it working. We will redo the permission/membership later on. Sounds good!

dsonck92 commented:
One question though, the migration removes the table altogether, doesn't that break ownership of resources that already exist?

bouassaba commented:
@dsonck92 no the tables "group_user" and "organization_user" were just redundant on top the real permission tables "userpermission" and "grouppermission".

I try to illustrate it in an example:
- When a user creates a group, (1) "owner" permission is inserted in "userpermission"
- When a guest is added to an existing group, (2) "viewer" permission is inserted in "userpermission" and (3) a membership entry is inserted in "group_user"

So as you can see (1) and (2) already define membership, naturally, (3) is just redundant, that's why I simplified it by removing the tables associated with (3).

bouassaba commented:
@dsonck92 so the short answer is "no" :) existing memberships are not broken, they are already defined by the core permission tables "userpermission" and "grouppermission".

While fixing this bug, I made sure to not break compatibility with existing installations, to be fully backward compatible.
